File: FE/web-app/src/types/voucher.ts
import { BasePaging } from "./api"

export enum DiscountType {
    Percentage = 'Percentage',
    Amount = 'Amount'
}

export enum VoucherStatus {
    Created = 'Created',   
    Effective = 'Effective',
    Expired = 'Expired'
}

export interface VoucherDetailDTO {
    id: string,
    name: string,
    code: string,
    discountType: DiscountType,
    status: VoucherStatus,
    discount: number,
    maxDiscount: number,
    quantity: number|null,
    isActive: boolean,
    remainingQuantity: number|null,
    validFrom: Date,
    validUntil: Date
}

export interface VoucherListItemDTO {
    id: string,
    name: string,
}

export interface CreateVoucherDTO {
    name: string,
    discountType: DiscountType,
    discount: number,
    maxDiscount: number,
    quantity: number|null,
    validFrom: string,
    duration: string
}

export interface SearchVoucherDTO extends BasePaging {
    name: string|null,
    code: string|null,
    start: Date|null,
    end: Date|null
}

File: BE/Core/Entities/VoucherFeature/Voucher.cs
using Core.Entities.PaymentFeature;

namespace Core.Entities.VoucherFeature
{
    public class Voucher
    {
        public string Id { get; set; }
        public string Name { get; set; }
        public string Code { get; set; }
        public DiscountType DiscountType { get; set; }
        public decimal Discount { get; set; }
        public decimal MaxDiscount { get; set; }
        public int? Quantity { get; set; }
        public DateTime ValidFrom { get; set; }
        public DateTime ValidUntil { get; set; }
        public bool IsActive { get; set; }
        public List<Order> AppliedOrders { get; set; } = new List<Order>();
    }
}
